pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java

```diff
@@ -256,4 +262,66 @@ private Map<String, String> consolidateMetadata(Map<String, String> customProper
   public void cancel(long requestId) {
     _opChainScheduler.cancel(requestId);
   }
+
+  public StagePlan explainQuery(
+      WorkerMetadata workerMetadata, StagePlan stagePlan, Map<String, String> requestMetadata) {
+
+    if (!workerMetadata.isLeafStageWorker()) {
+      LOGGER.debug("Explain query on intermediate stages is a NOOP");
+      return stagePlan;
+    }
+    long requestId = Long.parseLong(requestMetadata.get(CommonConstants.Query.Request.MetadataKeys.REQUEST_ID));
+    long timeoutMs = Long.parseLong(requestMetadata.get(CommonConstants.Broker.Request.QueryOptionKey.TIMEOUT_MS));
+    long deadlineMs = System.currentTimeMillis() + timeoutMs;
+
+    StageMetadata stageMetadata = stagePlan.getStageMetadata();
+    Map<String, String> opChainMetadata = consolidateMetadata(stageMetadata.getCustomProperties(), requestMetadata);
+
+    if (PipelineBreakerExecutor.hasPipelineBreakers(stagePlan)) {
+      // TODO: Support pipeline breakers before merging this feature.
+      LOGGER.error("Pipeline breaker is not supported in explain query");
+      return stagePlan;
+    }
```

Comment:
I guess we're not going to support queries with pipeline breakers with this new explain plan initially right? What do we want to show in those cases?

Reply:
I modified the description yesterday to indicate the PipelineBreaker state. Basically we in the current explain we just prune these nodes. In the new explain we will fail with that error shown here.

We can work on that feature in the future.

Reply:
The main problem is that in order to have the exact physical plan in this case we would need to actually execute the pipeline breaker part. For example, a query like:

```sql
select whatever from table1 
where col1 in (select something from table2 where col2 = cte)
```

the actual physical plan will depend on the result of `(select something from table2 where col2 = cte)`. Assuming that subquery is evaluated to [100, 200, 300], the query would be:

```sql
select whatever from table1 
where col1 in (100, 200, 300)
```

Which could, for example, use a inverted index. I had the happy idea of generating a random set of values on the `col1` type, but that could create incorrect plans. For example imagine we randomly generate the set of values `1,2`. The query we would be using to explain would be:

```sql
select whatever from table1 
where col1 in (1, 2)
```

Now imagine that values of `col1` went from 100 to 200. That query would be explained as `ALL_SEGMENTS_PRUNED_ON_SERVER` when the actual plan would be probably different

Reply:
Ah, I hadn't thought of that either, thanks for the explanation! I guess we can just update that TODO comment for now, it makes sense to defer this considering the current explain also doesn't properly support pipeline breaker.